src/libfsm/endids_internal.h

```diff
@@ -0,0 +1,42 @@
+#ifndef ENDIDS_INTERNAL_H
+#define ENDIDS_INTERNAL_H
+
+#include <stdlib.h>
+#include <stdint.h>
+
+#include <fsm/alloc.h>
+#include <fsm/capture.h>
+#include <fsm/fsm.h>
+#include <fsm/pred.h>
+
+#include <adt/stateset.h>
+
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+
+#include "internal.h"
+#include "endids.h"
+
+#define BUCKET_NO_STATE ((fsm_state_t)-1)
+#define DEF_BUCKET_COUNT 4
+#define DEF_BUCKET_ID_COUNT 16
+
+struct endid_info {
+	/* Add-only hash table, with a state ID and an associated
+	 * non-empty ordered array of unique end IDs. The state is the
+	 * key. Grows when the buckets are more than half full. */
+	unsigned bucket_count;
+	unsigned buckets_used;
```

Comment:
I suppose this is likely to get centralised under src/adt eventually. We have a couple of similar hash tables there already.

Reply:
I figured as much, but didn't want to do it prematurely. A dynamic hash table of type `state_id -> 'a`, but some of them are add-only (and much simpler for it), and they're internally storing different sizes of values, so it could get a bit messy.